File: 백준/Silver/17252. 삼삼한 수/삼삼한 수.py
"""
N 보다 같거나 작은 수들을 리스트에 모두 넣고

부분집합으로 모든 경우의 수를 체크해보기
"""

import sys

input = lambda: sys.stdin.readline().strip()

N = int(input())

lst = []
for i in range(21):
    temp = int(3 ** i)
    if temp <= N:
        lst.append(temp)

selected = [False] * len(lst)
possible = False


def subset(cur):
    global possible
    if possible:
        return
    if cur == len(lst):
        sum_value = 0
        for i in range(len(lst)):
            if selected[i]:
                sum_value += lst[i]
        if sum_value == N:
            possible = True
        return
    selected[cur] = True
    subset(cur + 1)
    selected[cur] = False
    subset(cur + 1)


subset(0)

print("YES" if N != 0 and possible else "NO")

File: 백준/Silver/17252. 삼삼한 수/삼삼한 수.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	private static int size;
	private static boolean[] selected;
	private static int[] lst;
	private static boolean possible;
	private static int N;
	
	public static void main(String[] args) throws NumberFormatException, IOException {
		N = Integer.parseInt(br.readLine());
		
		lst = new int[20];
		
		for (int i = 0; i < 20; i++) {
			int temp = (int) Math.pow(3, i);
			if (temp <= N) {
				size = i + 1;
				lst[i] = temp;				
			}
		}
		
		selected = new boolean[size];
		possible = false;
		
		subset(0);
		
		System.out.println(N != 0 && possible ? "YES" : "NO");
		
	} // end of main

	private static void subset(int cur) {
		if (possible) {
			return;
		}
		if (cur == size) {
			int sum = 0;
			for (int i = 0; i < size; i++) {
				if (selected[i]) {
					sum += lst[i];
				}
			}
			if (sum == N) {
				possible = true;
			}
			return;
		}
		selected[cur] = true;
		subset(cur + 1);
		selected[cur] = false;
		subset(cur + 1);
	} // end of subset
} // end of class